httpclient5/src/main/java/org/apache/hc/client5/http/impl/io/PoolingHttpClientConnectionManager.java

```diff
@@ -329,7 +331,22 @@ public synchronized ConnectionEndpoint get(
 
             @Override
             public boolean cancel() {
-                return leaseFuture.cancel(true);
+                if (cancelled.compareAndSet(false, true)) {
+                    final boolean result = leaseFuture.cancel(true);
+                    if (!result) {
+                        // The lease request has already completed successfully and cannot be cancelled
+                        try {
+                            final PoolEntry<HttpRoute, ManagedHttpClientConnection> poolEntry = leaseFuture.get(1, TimeUnit.MILLISECONDS);
```

Comment:
I'd prefer to see `leaseFuture.get(0, TimeUnit.MILLISECONDS)` here, by way of asserting that this call should not block

Reply:
@rschmitt This call should never block given at this point that the future is complete but you are absolutely right. Will do so.